project-templates/java-emf-theia/glsp-server/src/main/java/org/eclipse/glsp/example/javaemf/emf/EMFModelIndex.java

```diff
@@ -0,0 +1,271 @@
+/*******************************************************************************
+ * Copyright (c) 2022 EclipseSource and others.
+ *
+ * This program and the accompanying materials are made available under the
+ * terms of the Eclipse Public License v. 2.0 which is available at
+ * https://www.eclipse.org/legal/epl-2.0.
+ *
+ * This Source Code may also be made available under the following Secondary
+ * Licenses when the conditions for such availability set forth in the Eclipse
+ * Public License v. 2.0 are satisfied: GNU General Public License, version 2
+ * with the GNU Classpath Exception which is available at
+ * https://www.gnu.org/software/classpath/license.html.
+ *
+ * SPDX-License-Identifier: EPL-2.0 OR GPL-2.0 WITH Classpath-exception-2.0
+ ******************************************************************************/
+package org.eclipse.glsp.example.javaemf.emf;
+
+import java.util.ArrayList;
+import java.util.List;
+import java.util.Optional;
+import java.util.stream.Collectors;
+
+import com.google.common.base.Preconditions;
+import com.google.common.collect.BiMap;
+import com.google.common.collect.HashBiMap;
+
+import org.eclipse.emf.ecore.EObject;
+import org.eclipse.emf.ecore.util.EcoreUtil;
+import org.eclipse.emfcloud.modelserver.glsp.notation.Diagram;
+import org.eclipse.emfcloud.modelserver.glsp.notation.Edge;
+import org.eclipse.emfcloud.modelserver.glsp.notation.NotationElement;
+import org.eclipse.emfcloud.modelserver.glsp.notation.NotationFactory;
+import org.eclipse.emfcloud.modelserver.glsp.notation.SemanticElementReference;
+import org.eclipse.emfcloud.modelserver.glsp.notation.Shape;
+import org.eclipse.glsp.graph.GEdge;
+import org.eclipse.glsp.graph.GModelElement;
+import org.eclipse.glsp.graph.GModelRoot;
+import org.eclipse.glsp.graph.GNode;
+import org.eclipse.glsp.graph.GPoint;
+import org.eclipse.glsp.graph.GShapeElement;
+import org.eclipse.glsp.graph.impl.GModelIndexImpl;
+import org.eclipse.glsp.graph.util.GraphUtil;
+
+/**
+ * Is used to index all child elements of a semantic and notation model.
+ * Offers a set of query methods retrieve indexed elements.
+ */
+public final class EMFModelIndex extends GModelIndexImpl {
```

Comment:
I think it'd be great if there would be two index implementations:
1. The `EMFModelIndex` which maps GModel elements to `EObjects` (this one doesn't make any assumption on whether it is a notation or semantic element and can also be used if there is no separation.
2. A `EMFNotationModelIndex` that maps Semantic objects to Notation objects. This could be an extension of the `EMFModelIndex`

Reply:
Done with the current commit